src/spinnaker.cpp

```diff
@@ -0,0 +1,1049 @@
+    // TriggerSelectorEnums, so do not enable it as an input trigger.
+    meta->digital_lines = {
+        .line_count = 3,
+        .names = {
+          "Line0",
+          "Line1",
+          "Software",
+        },
+    };
+    meta->triggers = {
+        .exposure = { .input = 0, .output = 0b0010 },
+        .frame_start = { .input = 0b0101, .output = 0 },
+    };
+}
+
+void
+SpinnakerCamera::get_shape(struct ImageShape* shape) const
+{
+    const std::scoped_lock lock(lock_);
+
+    const uint32_t width = (int32_t)camera_->Width();
+    const uint32_t height = (int32_t)camera_->Height();
+
+    *shape = {
+        .dims = {
+            .channels = 1,
+            .width = width,
+            .height = height,
+            .planes = 1,
+        },
+        .strides = {
+          .channels = 1,
+          .width = 1,
+          .height = width,
+          .planes = width*height,
+        },
+        .type = to_sample_type(*(camera_->PixelFormat)),
+    };
+}
+
+void
+SpinnakerCamera::start()
+{
+    const std::scoped_lock lock(lock_);
+    frame_id_ = 0;
+    set_enum_node(camera_->AcquisitionMode, genicam_continuous);
+    camera_->BeginAcquisition();
+    started_ = true;
+}
+
+void
+SpinnakerCamera::stop()
+{
+    const std::scoped_lock lock(lock_);
+    // Disable the current trigger to prevent an effective deadlock between
+    // EndAcquisition and GetNextFrame that is awaiting a trigger.
+    if (IsWritable(camera_->TriggerMode)) {
+        camera_->TriggerMode = genicam_off;
+    }
+    // Could possibly use camera_->IsStreaming instead, but the Spinnaker
+    // docs are not clear enough about what that means. It's critical
+    // that we call EndAcquisiton exactly when needed so that associated
+    // buffers are freed.
+    if (started_) {
+        camera_->EndAcquisition();
+        started_ = false;
+    }
+}
+
+void
+SpinnakerCamera::execute_trigger() const
+{
+    const std::scoped_lock lock(lock_);
+    camera_->TriggerSoftware();
+}
+
+void
+SpinnakerCamera::get_frame(void* im, size_t* nbytes, struct ImageInfo* info)
+{
+    // Adapted from the Acquisition.cpp example distributed with the Spinnaker
+    // SDK. Cannot acquire the lock here because GetNextImage may await a
+    // trigger indefinitely effectively causing a deadlock with other methods
+    // that acquire the camera lock (like stop).
+    Spinnaker::ImagePtr frame = camera_->GetNextImage();
+
+    // Acquire the lock once we have the frame so that we can check if the
+    // camera was stopped while we were waiting. In the case that stop was
+    // called before we got here, the frame is not valid and its buffer backed
+    // memory has been released, so return without explicitly releasing.
+    const std::scoped_lock lock(lock_);
+    if (!started_) {
+        return;
+    }
+
+    if (frame->IsIncomplete()) {
+        LOGE(
+          "Image incomplete: %s",
+          Spinnaker::Image::GetImageStatusDescription(frame->GetImageStatus()));
```

Comment:
Might be nice to know the frame id here.

```suggestion
        LOGE(
          "Image incomplete: (%d) %s",
          (int)frame->GetFrameID(),
          Spinnaker::Image::GetImageStatusDescription(frame->GetImageStatus()));
```